lucene/core/src/java/org/apache/lucene/index/StandardDirectoryReader.java

```diff
@@ -39,33 +40,44 @@
 
   final IndexWriter writer;
   final SegmentInfos segmentInfos;
+  private final Comparator<LeafReader> leafSorter;
   private final boolean applyAllDeletes;
   private final boolean writeAllDeletes;
 
-  /** called only from static open() methods */
+  /**
+   * package private constructor, called only from static open() methods. If parameter {@code
+   * leafSorter} is not {@code null}, the provided {@code readers} are expected to be already sorted
```

Comment:
Could we add an `assertLeavesSorted` method to confirm they really were properly sorted by the caller?  There are so many paths for creating `StandardDirectoryReader` through are code base that it would not surprise me if one of the misses to sort properly.

Reply:
Great suggestion! I moved sorting of leaves to the constructor. 
Addressed in 8155f2e2abd4ed3a6280aae940d018dfe3b3dad1.